modules/core/src/main/java/org/apache/ignite/internal/jdbc/thin/JdbcThinConnection.java

```diff
@@ -1659,45 +1747,155 @@ private void connectInAffinityAwarenessMode(HostAndPortRange[] srvs) throws SQLE
 
-            // Partition result was requested.
-            if (res.response() instanceof JdbcQueryExecuteResult && qryReq.partitionResponseRequest()) {
-                PartitionResult partRes = ((JdbcQueryExecuteResult)res.response()).partitionResult();
+        /** {@inheritDoc} */
+        @Override public void run() {
+            try {
+                for (Map.Entry<InetSocketAddress, Integer> delayEntry : reconnectionDelaysRemainder.entrySet())
+                    reconnectionDelaysRemainder.put(delayEntry.getKey(), delayEntry.getValue() - RECONNECTION_DELAY);
 
-                if (partRes == null || affinityCache.version().equals(partRes.topologyVersion())) {
-                    int cacheId = (partRes != null && partRes.tree() != null) ?
-                        GridCacheUtils.cacheId(partRes.cacheName()) :
-                        -1;
+                Set<InetSocketAddress> aliveSockAddrs =
+                    ios.values().stream().map(JdbcThinTcpIo::socketAddress).collect(Collectors.toSet());
 
-                    PartitionClientContext partClientCtx = partRes != null ?
-                        new PartitionClientContext(partRes.partitionsCount()) :
-                        null;
+                IgniteProductVersion prevIgniteEndpointVer = null;
 
-                    QualifiedSQLQuery qry = new QualifiedSQLQuery(qryReq.schemaName(), qryReq.sqlQuery());
+                for (int i = 0; i < connProps.getAddresses().length; i++) {
+                    HostAndPortRange srv = connProps.getAddresses()[i];
 
-                    JdbcThinPartitionResultDescriptor partResDescr =
-                        new JdbcThinPartitionResultDescriptor(partRes, cacheId, partClientCtx);
+                    try {
+                        InetAddress[] addrs = InetAddress.getAllByName(srv.host());
 
-                    affinityCache.addSqlQuery(qry, partResDescr);
+                        for (InetAddress addr : addrs) {
+                            for (int port = srv.portFrom(); port <= srv.portTo(); ++port) {
+                                InetSocketAddress sockAddr = null;
+
+                                try {
+                                    sockAddr = new InetSocketAddress(addr, port);
+
+                                    if (aliveSockAddrs.contains(sockAddr)) {
+                                        reconnectionDelaysRemainder.remove(sockAddr);
+                                        reconnectionDelays.remove(sockAddr);
+
+                                        continue;
+                                    }
+
+                                    Integer delayRemainder = reconnectionDelaysRemainder.get(sockAddr);
+
+                                    if (delayRemainder != null && delayRemainder != 0)
+                                        continue;
+
+                                    if (closed) {
+                                        maintenanceExecutor.shutdown();
```

Comment:
Why do we need to call `shutdown` here?

Reply:
We should prevent rerunning reconnection task if connection itself is in closing state, so we have a choice either call `connectionsHndScheduledFut.cancel()` or `maintenanceExecutor.shutdown()`. Both options are fine here. The only very minor point to favor shutdown() to cancel is the fact that it will also stop `qryTimeoutScheduledFut` a little bit earlier.